### PR TITLE
Wrap connections in an N minute timeout to ensure they get reaped correctly

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/http/endpoint.py [hunk 1]
```diff
@@ -66,13 +66,63 @@ def matrix_federation_endpoint(reactor, destination, ssl_context_factory=None,
...
+    def _time_things_out_maybe(self):
+        if time.time() - self.last_request >= 2 * 60:
+            self.abort()
+
+    def request(self, request):
+        self.last_request = time.time()
+
+        # Time this connection out if we haven't send a request in the last
+        # N minutes
+        reactor.callLater(3 * 60, self._time_things_out_maybe)
```
Comment (ara4n) [inline synapse/http/endpoint.py]: why is it 3 mins here, but 2 mins in the _time_things_out_maybe above?

File: synapse/http/endpoint.py [hunk 2]
```diff
@@ -66,13 +66,63 @@ def matrix_federation_endpoint(reactor, destination, ssl_context_factory=None,
...
             default_port=default_port, endpoint=transport_endpoint,
             endpoint_kw_args=endpoint_kw_args
-        )
+        ))
     else:
-        return transport_endpoint(reactor, domain, port, **endpoint_kw_args)
+        return _WrappingEndointFac(transport_endpoint(reactor, domain, port, **endpoint_kw_args))
+
+
+class _WrappingEndointFac(object):
```
Comment (ara4n) [inline synapse/http/endpoint.py]: s/Endoint/Endpoint/ probably ;)

File: synapse/http/matrixfederationclient.py [hunk 3]
```diff
@@ -61,6 +61,11 @@
 MAX_SHORT_RETRIES = 3
 
 
+def test(conn):
```
Comment (ara4n) [inline synapse/http/matrixfederationclient.py]: hm, is this meant to be here?

File: synapse/http/endpoint.py [hunk 4]
```diff
@@ -66,13 +66,67 @@ def matrix_federation_endpoint(reactor, destination, ssl_context_factory=None,
...
+        reactor.callLater(3 * 60, self._time_things_out_maybe)
+
+        d = self.conn.request(request)
+
+        def update_request_time(res):
+            self.last_request = time.time()
+            reactor.callLater(3 * 60, self._time_things_out_maybe)
+            return res
+
+        d.addCallback(update_request_time)
```
Comment (ara4n) [inline synapse/http/endpoint.py]: why do we do this by both a callback and a reactor.callLater()?

Comment (erikjohnston) [inline synapse/http/endpoint.py]: We schedule a timeout in 3 minutes both before we send the request, and after we received the response.

File: synapse/http/endpoint.py [hunk 5]
```diff
@@ -66,13 +66,67 @@ def matrix_federation_endpoint(reactor, destination, ssl_context_factory=None,
...
+
+    @defer.inlineCallbacks
+    def connect(self, protocolFactory):
+        conn = yield self.endpoint_fac.connect(protocolFactory)
+        conn = _WrappedConnection(conn)
+        defer.returnValue(conn)
+
+
+class _WrappedConnection(object):
+    """Wraps a connection and calls abort on it if it hasn't seen any actio
```
Comment (uhoreg) [inline synapse/http/endpoint.py]: missing "n" in "action".  Also, the comment says 5 minutes, but the code says 3 minutes?

**Pull request merged**